File: c4proto-util/src/main/scala/ee/cone/c4proto/AssemblerGen.scala

package ee.cone.c4proto
import Types._

abstract class Join1[T1, R,TK,RK](
  t1: WorldKey[Index[TK,T1]], 
  val outputWorldKey: WorldKey[Index[RK,R]]
) extends Join[R,TK,RK] {
  def join(a1: Values[T1]): Values[(RK,R)]
  //
  def joins(in: Seq[Values[Object]]): Iterable[(RK,R)] = in match {
    case Seq(a1) ⇒
      join(a1.asInstanceOf[Values[T1]])
  }
  def inputWorldKeys: Seq[WorldKey[Index[TK,Object]]] =
    Seq(t1).asInstanceOf[Seq[WorldKey[Index[TK,Object]]]]
}

abstract class Join2[T1, T2, R,TK,RK](
  t1: WorldKey[Index[TK,T1]], t2: WorldKey[Index[TK,T2]], 
  val outputWorldKey: WorldKey[Index[RK,R]]
) extends Join[R,TK,RK] {
  def join(a1: Values[T1], a2: Values[T2]): Values[(RK,R)]
  //
  def joins(in: Seq[Values[Object]]): Iterable[(RK,R)] = in match {
    case Seq(a1, a2) ⇒
      join(a1.asInstanceOf[Values[T1]], a2.asInstanceOf[Values[T2]])
  }
  def inputWorldKeys: Seq[WorldKey[Index[TK,Object]]] =
    Seq(t1, t2).asInstanceOf[Seq[WorldKey[Index[TK,Object]]]]
}

abstract class Join3[T1, T2, T3, R,TK,RK](
  t1: WorldKey[Index[TK,T1]], t2: WorldKey[Index[TK,T2]], t3: WorldKey[Index[TK,T3]], 
  val outputWorldKey: WorldKey[Index[RK,R]]
) extends Join[R,TK,RK] {
  def join(a1: Values[T1], a2: Values[T2], a3: Values[T3]): Values[(RK,R)]
  //
  def joins(in: Seq[Values[Object]]): Iterable[(RK,R)] = in match {
    case Seq(a1, a2, a3) ⇒
      join(a1.asInstanceOf[Values[T1]], a2.asInstanceOf[Values[T2]], a3.asInstanceOf[Values[T3]])
  }
  def inputWorldKeys: Seq[WorldKey[Index[TK,Object]]] =
    Seq(t1, t2, t3).asInstanceOf[Seq[WorldKey[Index[TK,Object]]]]
}

File: c4actor-base/src/main/scala/ee/cone/c4actor/AssemblerGen.pl

use strict;
use warnings FATAL => 'all';

open FF, '>', "AssemblerGen.scala" or die;
print FF "
package ee.cone.c4actor
import Types._
";
for my $n(1..3) {
    my $jm = sub { my ($m,$c) = @_; my $r = join ', ', map{&$m()} 1..$n; $r ? \"$r$c" : \$r };
    print FF qq^
abstract class Join$n\[${&$jm(sub{"T$_"},', ')}R,TK,RK](
  ${&$jm(sub{"t$_: WorldKey[Index[TK,T$_]]"},', ')}
  val outputWorldKey: WorldKey[Index[RK,R]]
) extends Join[R,TK,RK] {
  def join(${&$jm(sub{"a$_: Values[T$_]"},'')}): Values[(RK,R)]
  //
  def joins(in: Seq[Values[Object]]): Iterable[(RK,R)] = in match {
    case Seq(${&$jm(sub{"a$_"},'')}) ⇒
      join(${&$jm(sub{"a$_.asInstanceOf[Values[T$_]]"},'')})
  }
  def inputWorldKeys: Seq[WorldKey[Index[TK,Object]]] =
    Seq(${&$jm(sub{"t$_"},'')}).asInstanceOf[Seq[WorldKey[Index[TK,Object]]]]
}
^
}
close FF or die;
1;